### DISCUSSION TITLE
prepare_freethreaded_python crashes in PyOS_InputHook

### DISCUSSION THREAD

**Discussion by chubei-oppen** [PyO3/pyo3#1986] [General]
I have an extension module which sets [`PyOS_InputHook`](https://docs.python.org/3.6/c-api/veryhigh.html#c.PyOS_InputHook). Calling [`prepare_freethreaded_python`](https://pyo3.rs/main/doc/pyo3/fn.prepare_freethreaded_python.html) in the hook will crash the process with `Fatal Python error: take_gil: NULL tstate`. Calling it on other sites is fine.

I'm not an expert with `Cython` and not sure if this is a `Python` problem or `pyo3` one. Any input is appreciated.

Environment:

Ubuntu 18.04.6 LTS
Python 3.6.13 :: Anaconda, Inc.
rustc 1.58.0-nightly (29b124802 2021-10-25)
pyo3 0.15.0 features = ["extension-module"]

**Comment by davidhewitt** [comment]
Are you able to share minimal repro?

Python 3.6 is quite old now and works very differently with threads compared to Python 3.7+. Is it possible for you to upgrade to a newer Python version?

**Reply by chubei-oppen** [reply]
The minimal repo: https://github.com/chubei-oppen/free-thread-crash

Python 3.7 doesn't have the problem, so this is a Python 3.6 bug?

**Reply by davidhewitt** [reply]
Yes, I think so.

TBH in that case, as Python 3.6 is end-of-life next month and PyO3 is expecting to drop support for it soon after that, I'm unlikely to try to find time to fix this one.

If anyone else wants to contribute a fix, PyO3 0.15.1 might be the last PyO3 version to support 3.6; we could release 0.15.2 with this fix if needed.

**Reply by chubei-oppen** [reply]
Thanks. Lucky for me there's a workaround (calling prepare_freethreaded_python out of the hook). Guess I'll live with the bug for a while.
